### DISCUSSION TITLE
Why is only QFocal loss implemented and not GFocal loss?

### DISCUSSION THREAD

**Discussion by borschy** [ultralytics/yolov5#2747] [Q&A] [answered]
According to the paper proposing QFocal loss, []https://arxiv.org/pdf/2006.04388.pdf], Distribution Focal loss is also proposed in order to fuzzify otherwise inflexible representations of bounding boxes. Is there a reason only QFocal loss was chosen to be implemented and not Distribution Focal loss as well? @yxNONG

**Comment by glenn-jocher** [comment, marked as answer]
@borschy QFocal loss was implemented in PR https://github.com/ultralytics/yolov5/pull/1482 by @yxNONG, who might best be able to answer your question.

**Reply by borschy** [reply]
ok, thanks

**Reply by yxNONG** [reply]
@borschy Sorry for my late reply, 
to my understanding，The paper  [https://arxiv.org/pdf/2006.04388.pdf] use a distribution rather then a point to represent the bounding box,
for a single bbox, it need to regressed not [ 4 ] points but [ 4 * reg_max] points (noticed that here use #reg_max discrete point to approximate a continue distribution) 
Using the a complete version of this paper, the backbone of yolov5 must be changed 
however, i did not found a good way to switch the network yet

**Reply by borschy** [reply]
@yxNONG I see, thanks for the reply :D

**Comment by indigohana** [comment]
Hello, if qfocal loss wants to be applied to the confidence loss, according to my understanding, do you want to change the model.gr in train.py from the default 1.0 to iou? I look forward to receiving your reply, thank you!

**Reply by yxNONG** [reply]
@indigohana 
qfocal loss is more general version of focal loss(which design for target distribution is one-hot)， if the target is one-hot, then gfocal loss is the same as focal.
however, if applying to label smooth in cls target(not one-hot), gfocal loss is much more make sense for me
i not sure whether the default setting of target of conf loss is iou or 1
as for me, if focal loss wanted to apply in a loss which target is not one-hot, i will change focal loss to gfocal

**Reply by glenn-jocher** [reply]
@yxNONG @indigohana in YOLOv5 the default confidence target is equal to the IoU of the predictions, not 1.0. This produces better ordering of the boxes during NMS.

Every output from YOLOv5 and any other model may well be considered a distribution rather than a point, the current output is simply the location of the distribution maximum likelihood value, i.e. the mean in a Normal PDF. Unfortunately ML models have not typically focused on metric outputs that pair a max likelihood point with it's distribution. This would allow for better confidence bounds in some cases, like the bounding boxes as you mentioned.
